### PR TITLE
feat: Message Read Indicator

### LINKED ISSUE #2099: Display indicator if message was read by customer
**Is your enhancement request related to a problem? Please describe.**
After replying to a customer (usually on live chat widget channel), it's not clear if the customer saw the message or not. I'm asking myself if I should send an email too.
The only related indicator now is the green "online" badge on the customer icon, in the conversations list view.

**Describe the solution you'd like**
Please consider a solution similar to WhatsApp, showing an indicator on sent messages / received by client / read by client.
Where possible, this should work multi-channel, eg Live chat / WhatsApp / email via 1px img embed (if enabled)

### PR DESCRIPTION
**Pull request by fayazara**
## Description

This will allow agents to check if the user has read the message.
Screenshot2022-04-13 at 18 22 48

Demo - https://www.loom.com/share/0bfbe0575e954924bd25b356f564dddc

Fixes #2099 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/javascript/dashboard/components/widgets/conversation/Message.vue [hunk 1]
```diff
@@ -153,6 +154,10 @@ export default {
       type: Boolean,
       default: false,
     },
+    userReadMessage: {
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/Message.vue:157]: Does 'hasUserReadMessage' more descriptive?

File: app/javascript/dashboard/mixins/conversations.js [hunk 2]
```diff
@@ -15,6 +15,9 @@ export default {
           chat.private !== true
       ).length;
     },
+    userReadMessage(created_at, contact_last_seen) {
+      return !(contact_last_seen - created_at < 0);
```
Comment (nithindavid) [inline app/javascript/dashboard/mixins/conversations.js:19]: Please use camelCase here.

**Review by nithindavid — CHANGES_REQUESTED** [review]
Please see the Minor comments left

File: app/javascript/dashboard/components/widgets/conversation/Message.vue (hunk 1 above)
Comment (fayazara) [reply app/javascript/dashboard/components/widgets/conversation/Message.vue]: Resolved in https://github.com/chatwoot/chatwoot/pull/4224/commits/20d22f85305b8493d1c62277dedfbb7a1da2b613

**Comment by fayazara** [comment]
@pranavrajs @muhsin does this approach work for all inbox types? I am using `message.created_at` and `contact_last_seen` props to handle, do I need to handle the inbox type as well here?

**Comment by iamsivin** [comment]
@fayazara Found an issue, the tick mark is not showing properly. Can you check this screencast 
 
https://user-images.githubusercontent.com/64252451/160345753-91f62531-2e42-4d32-a735-56b573920f9b.mov

**Comment by fayazara** [comment]
> @fayazara Found an issue, the tick mark is not showing properly. Can you check this screencast

@pranavrajs @muhsin-k  The `contact_last_seen` is not udpated via a websocket, so the ticks are not getting udpated in realtime, any other way to handle this correctly?

**Comment by muhsin-k** [comment]
> > @fayazara Found an issue, the tick mark is not showing properly. Can you check this screencast
> 
> @pranavrajs @muhsin-k The `contact_last_seen` is not udpated via a websocket, so the ticks are not getting udpated in realtime, any other way to handle this correctly?

@sojan-official @pranavrajs  What about emitting `message_updated` event  after reading a message?

**Comment by sojan-official** [comment]
@muhsin-k @pranavrajs  ref: https://github.com/chatwoot/chatwoot/issues/2587
rather than doing this calculation based on contact_last_seen_at, what about introducing a `message_delivery_status` enum for messages 

1: send ( created in chatwoot database, default ), ( retry button for external channels if status doesn't turn to delivered even after x time )
2:delivered ( for external sources, when the source_id is updated ) ( for external channels they can mark this attribute to let agents know that the message was received at their end )
3. read ( double ticks, for supported channels )

we can use `message. updated` for this case.

**Comment by muhsin-k** [comment]
> @muhsin-k @pranavrajs ref: #2587 rather than doing this calculation based on contact_last_seen_at, what about introducing a `message_delivery_status` enum for messages
> 
> 1: send ( created in chatwoot database, default ), ( retry button for external channels if status doesn't turn to delivered even after x time ) 2:delivered ( for external sources, when the source_id is updated ) ( for external channels they can mark this attribute to let agents know that the message was received at their end ) 3. read ( double ticks, for supported channels )
> 
> we can use `message. updated` for this case.

That makes sense. This approach is more straightforward.

**Comment by fayazara** [comment]
> That makes sense. This approach is more straightforward.


Do we have this already?

**Comment by sojan-official** [comment]
@fayazara  nope, we don't have the APIs for this at the moment.

**Comment by sojan-official** [comment]
@muhsin-k  @fayazara `contact_last_seen` attribute is saved on `conversation`. so let's add a `conversation_updated` socket event and proceed with the ticket. let's take up adding the read status attribute in messages as a separate ticket.

File: app/javascript/dashboard/components/widgets/conversation/MessagesView.vue [hunk 3]
```diff
@@ -241,6 +250,10 @@ export default {
       }
       return 'arrow-chevron-left';
     },
+    getLastSeenAt() {
+      if (this.conversationLastSeen) return this.conversationLastSeen;
+      return this.currentChat.contact_last_seen_at;
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/MessagesView.vue:255]:
```suggestion
      const {contact_last_seen_at: contactLastSeenAt } = this.currentChat;
      return contactLastSeenAt;
```

File: app/javascript/dashboard/helper/actionCable.js [hunk 4]
```diff
@@ -64,6 +65,13 @@ class ActionCableConnector extends BaseActionCableConnector {
     this.fetchConversationStats();
   };
 
+  onConversationRead = data => {
+    this.app.$store.dispatch(
+      'updateConversationRead',
+      data.contact_last_seen_at
```
Comment (nithindavid) [inline app/javascript/dashboard/helper/actionCable.js:71]: Use camelKeys here also;

**Review by nithindavid — CHANGES_REQUESTED** [review]

File: app/javascript/dashboard/helper/actionCable.js (hunk 4 above)
Comment (fayazara) [reply app/javascript/dashboard/helper/actionCable.js]: Resolved

File: app/javascript/dashboard/components/widgets/conversation/MessagesView.vue (hunk 3 above)
Comment (fayazara) [reply app/javascript/dashboard/components/widgets/conversation/MessagesView.vue]: Resolved

File: app/javascript/dashboard/mixins/conversations.js (hunk 2 above)
Comment (fayazara) [reply app/javascript/dashboard/mixins/conversations.js]: resolved

**Review by muhsin-k — DISMISSED** [review]

**Review by iamsivin — DISMISSED** [review]
LGTM 👍🏻

File: app/javascript/dashboard/helper/actionCable.js [hunk 5]
```diff
@@ -64,6 +65,11 @@ class ActionCableConnector extends BaseActionCableConnector {
     this.fetchConversationStats();
   };
 
+  onConversationRead = data => {
+    const lastSeen = data.contact_last_seen_at;
```
Comment (muhsin-k) [inline app/javascript/dashboard/helper/actionCable.js:69]:
```suggestion
    const { contact_last_seen_at: lastSeen } = data;
```

**Review by nithindavid — DISMISSED** [review]

**Comment by iamsivin** [comment]
@fayazara Have a doubt, don't know is an issue or not, If the widget is open and when we sent and receive a message the message read indication is not working and if the widget is closed and we sent a message then it works as expected. You can see it in this screencast.

https://www.loom.com/share/6d6075cea60e490480413519352cbbd3

**Comment by fayazara** [comment]
> @fayazara Have a doubt, don't know is an issue or not, If the widget is open and when we sent and receive a message the message read indication is not working and if the widget is closed and we sent and receive a message then it works as expected. You can see it in this screencast. https://www.loom.com/share/6d6075cea60e490480413519352cbbd3

Looks like a bug, looking into it.

**Comment by fayazara** [comment]
@pranavrajs @muhsin-k The `conversation:read` is not triggered when the chat window is popped out.

Another thing I noticed is, the widget button stays even after opening the chat window (This used to close earlier, If I am not wrong). 

https://user-images.githubusercontent.com/15716057/163759579-42f2d185-ad37-4fdb-9eca-5a3145c4c832.mp4

**Comment by muhsin-k** [comment]
> @pranavrajs @muhsin-k The `conversation:read` is not getting triggered when the chat window is popped out.
> 
> Another thing I noticed is, the widget button stays even after opening the chat window (This used to close earlier, If I am not wrong).
> 
>  Screenshot2022-04-18.at.10.57.51.mp4

Seems like the `contact_last_seen ` is not updating in pop out window view. Let me look in to this

**Review by iamsivin — DISMISSED** [review]

**Review by pranavrajs — APPROVED** [review]

**Pull request merged**